### PR TITLE
Adds filter bar component

### PR DESCRIPTION
**Pull request by Delerme**
WIP, waiting on filter query endpoint to work to test E2E, but it's worth taking a look now

### CODE REVIEW

File: src/core/core.js [hunk 1]
```diff
@@ -59,9 +59,9 @@ export default class Core {
     });
   }
 
-  verticalSearch(queryString, verticalKey) {
+  verticalSearch(queryString, verticalKey, filter) {
```
Comment (b1lly) [inline src/core/core.js:62]: Note:

Usually when we start going more than 2 params in a signature, my spidey senses start tingling to start thinking about using an Object as the interface instead. Even if you don't do it now, it's worth starting to think about for the next iteration. Along that train of thinking, I also consider the difference between required arguments vs optional ones. Once theres a combination of those things, I tend to start itching for a refactor.

File: src/core/search/autocompletedatatransformer.js [hunk 2]
```diff
@@ -39,6 +39,29 @@ export default class AutoCompleteDataTransformer {
...
+      highlightedSections.push({
+        label: section.label,
+        results: highlightedResults,
+      });
+    }
+
+    return AutoCompleteDataTransformer.clean('autocomplete.input', {
+      'sections': highlightedSections
+    });
+  }
```
Comment (b1lly) [inline src/core/search/autocompletedatatransformer.js]: We might want to start thinking about a view model here, similar to what I started for the search transformers.

File: src/core/search/search.js [hunk 3]
```diff
@@ -13,7 +13,7 @@ export default class Search {
      * @type {string}
      * @private
      */
-    this._baseUrl = this._isLocal ? 'http://' + window.location.hostname : 'https://liveapi.yext.com';
+    this._baseUrl = !isLocal ? 'http://' + window.location.hostname : 'https://liveapi.yext.com';
```
Comment (b1lly) [inline src/core/search/search.js]: Is this logic right?

File: src/core/search/autocompletedatatransformer.js [hunk 4]
```diff
@@ -39,6 +39,29 @@ export default class AutoCompleteDataTransformer {
...
+          result,
+          AutoCompleteDataTransformer.highlight(value, subStrings)));
+      }
+      highlightedSections.push({
+        label: section.label,
+        results: highlightedResults,
+      });
+    }
+
+    return AutoCompleteDataTransformer.clean('autocomplete.input', {
```
Comment (b1lly) [inline src/core/search/autocompletedatatransformer.js]: Shouldn't the moduleId be dynamic? You might need to bring in the bar key into the transformation as input. The reason being, if you were to append multiple 'FilterPicker' components on the page, the data should be scoped to the FilterPicker bar key provided.

Originally, that was the "quick fix", if you know of a better way that would be great too.

File: src/core/search/autocompletedatatransformer.js [hunk 5]
```diff
@@ -39,6 +39,29 @@ export default class AutoCompleteDataTransformer {
...
+
+        highlightedResults.push(Object.assign(
+          result,
+          AutoCompleteDataTransformer.highlight(value, subStrings)));
+      }
+      highlightedSections.push({
+        label: section.label,
+        results: highlightedResults,
+      });
+    }
```
Comment (b1lly) [inline src/core/search/autocompletedatatransformer.js]: It might be worth creating an actual model for the FilterPicker data. I know it will probably be simple to start, but what do you think about establishing a practice that we have concrete models for all of this data that comes back.

File: src/ui/components/search/autocompletecomponent.js [hunk 6]
```diff
@@ -25,6 +25,11 @@ export default class AutoCompleteComponent extends Component {
   constructor(opts = {}) {
     super(opts)
 
+    /**
+     * Whether autocomplete is simple or filter
+     */
+    this.filterSearch = opts.filterSearch || false
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js]: isFilterPicker? Not sure I like this name.

File: src/ui/components/search/autocompletecomponent.js [hunk 7]
```diff
@@ -25,6 +25,11 @@ export default class AutoCompleteComponent extends Component {
   constructor(opts = {}) {
     super(opts)
 
+    /**
+     * Whether autocomplete is simple or filter
+     */
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js:31]: @type {boolean}

File: src/ui/components/search/autocompletecomponent.js [hunk 8]
```diff
@@ -290,5 +303,17 @@ export default class AutoCompleteComponent extends Component {
...
+    if (key === Keys.ENTER) {
+      e.preventDefault();
+
+      const filter = JSON.stringify(sections[this._sectionIndex].results[this._resultIndex].filter);
+
+      this.updateQuery(value);
+      this._onSubmit(value, filter);
+      this.reset();
+      this.close();
+    }
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js:327]: I'm not sure we want the keyboard navigation to handle submission. You can remove this here.

File: src/ui/components/search/autocompletecomponent.js [hunk 9]
```diff
@@ -242,10 +251,14 @@ export default class AutoCompleteComponent extends Component {
...
-    this.core.autoComplete(value, this._experienceKey, this._barKey);
+    if (this.filterSearch) {
+      this.core.autoCompleteFilter(value, this._experienceKey, this._barKey);
+    } else {
+      this.core.autoComplete(value, this._experienceKey, this._barKey);
+    }
   }
 
-  handleNavigateResults(key, e) {
+  handleNavigateResults(key, value, e) {
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js]: Navigation doesn't need to handle submission so you can remove this as an input arg.

File: src/ui/components/search/autocompletecomponent.js [hunk 10]
```diff
@@ -200,7 +209,7 @@ export default class AutoCompleteComponent extends Component {
     if (opt_value === undefined) {
       let sections = this._state.get('sections'),
           results = sections[this._sectionIndex].results;
-      opt_value = results[this._resultIndex].shortValue;
+      opt_value = results[this._resultIndex].shortValue || results[this._resultIndex].value;
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js]: The model (or state) thats provided to this component should handle this type of logic. We should make the autocomplete operate off one model. I mentioned creating a concrete model in the data transformation layer; if we did that, this or case would be avoided for instance.

File: src/ui/components/search/autocompletecomponent.js [hunk 11]
```diff
@@ -145,22 +150,26 @@ export default class AutoCompleteComponent extends Component {
...
     DOM.on(queryInput, 'keydown', (e) => {
-     this.handleNavigateResults(e.keyCode, e);
+      this.handleNavigateResults(e.keyCode, queryInput.value, e);
     });
 
     // Allow the user to select a result with the mouse
     DOM.delegate(this._container, '.js-yext-autocomplete-option', 'mousedown', (evt, target) => {
       let data = target.dataset,
-          val = data.value,
+          val = data.short || data.value,
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js]: See model comment below.

File: src/ui/components/search/filterpicker.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
+import Component from '../component';
+import DOM from '../../dom/dom';
+
+export default class FilterPicker extends Component {
```
Comment (b1lly) [inline src/ui/components/search/filterpicker.js:4]: Do you think its worth having a common base class that this extends?

File: src/ui/components/search/filterpicker.js [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+     * @type {string} CSS selector
+     */
+    this._inputEl = opts.inputEl || '.js-yext-query';
+
+    /**
+     * The title used, provided to the template as a data point
+     * Optionally provided.
+     * @type {string}
+     */
+    this.title = opts.title || 'Answers Universal Search';
```
Comment (b1lly) [inline src/ui/components/search/filterpicker.js]: This probably isn't needed.

File: src/ui/components/search/filterpicker.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    /**
+     * The filter string to use for the provided query
+     * Optionally provided
+     * @type {string}
+     */
+    this.filter = opts.filter || this.getUrlParams().get('filter') || '';
+  }
+
+  static get type() {
+    return 'FilterBar';
```
Comment (b1lly) [inline src/ui/components/search/filterpicker.js]: I think FilterPicker is probably a fine name. Might not need to override this.

**Review by b1lly — COMMENTED** [review]
Main comments are around creating a concrete data model in the data transformation layer, for the FilterPicker component / AutoComplete, and stylistic nits. Otherwise nice job!

**Review by b1lly — COMMENTED** [review]
Main comments are around creating a concrete data model in the data transformation layer, for the FilterPicker component / AutoComplete, and stylistic nits. Otherwise nice job!

File: src/core/core.js (hunk 1 above)
Comment (Delerme) [reply src/core/core.js:62]: Yea as we start refactoring core I think it makes sense to have immutable request objects with required and optional keys. Will wait until refactor for this case

File: src/core/search/autocompletedatatransformer.js (hunk 5 above)
Comment (Delerme) [reply src/core/search/autocompletedatatransformer.js]: Done

File: src/core/search/autocompletedatatransformer.js (hunk 4 above)
Comment (Delerme) [reply src/core/search/autocompletedatatransformer.js]: Yup, done

File: src/core/search/autocompletedatatransformer.js (hunk 2 above)
Comment (Delerme) [reply src/core/search/autocompletedatatransformer.js]: Done

File: src/core/search/search.js (hunk 3 above)
Comment (Delerme) [reply src/core/search/search.js]: No, I was using it for testing so I didn't have to run the api locally. Fixed

File: src/ui/components/search/autocompletecomponent.js (hunk 7 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js:31]: done

File: src/ui/components/search/autocompletecomponent.js (hunk 6 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: changed

File: src/ui/components/search/autocompletecomponent.js (hunk 11 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: this case was actually to fix SearchBar, for whatever reason value is no longer set with the old autocomplete endpoint. Will revisit this later, as I think SearchBar needs a data model now

File: src/ui/components/search/autocompletecomponent.js (hunk 10 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: Done

File: src/ui/components/search/filterpicker.js (hunk 13 above)
Comment (Delerme) [reply src/ui/components/search/filterpicker.js]: Removed

File: src/ui/components/search/filterpicker.js (hunk 14 above)
Comment (Delerme) [reply src/ui/components/search/filterpicker.js]: It seems a little confusing, since a picker is generally a single or multi select dropdown component. FilterBar sound like SearchBar which I think is good because they are very similar behavior-wise

File: src/ui/components/search/filterpicker.js (hunk 12 above)
Comment (Delerme) [reply src/ui/components/search/filterpicker.js:4]: Not sure, it doesn't seem like we'd get a whole lot from it, most of the shared code here is just boilerplate that many components have, so there's probably a better solution than inheritance (compose a config object or something?)

File: src/ui/components/search/autocompletecomponent.js (hunk 9 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: I think it makes more sense, otherwise it's super confusing as to why enter even works. Also, this function has access to the data needed whereas putting it in filterpicker.js would require changing the form to send the necessary data.

File: src/ui/components/search/autocompletecomponent.js (hunk 8 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js:327]: As above, I think submission on enter makes more sense here than anywhere else

File: src/core/models/filtersdata.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
+export default class FiltersData {
```
Comment (b1lly) [inline src/core/models/filtersdata.js]: Is this FiltersData or AutoCompleteData? 'Cause isn't the highlighting chunk also used for the normal auto complete that isn't Filters?

File: src/ui/components/search/filterpicker.js (hunk 14 above)
Comment (b1lly) [reply src/ui/components/search/filterpicker.js]: It's possible that we'll expand it into both single and multi in the future.

File: src/ui/components/search/autocompletecomponent.js (hunk 8 above)
Comment (b1lly) [reply src/ui/components/search/autocompletecomponent.js:327]: If that's the case, I still think there should be a separate handler for submission. It shouldn't be part of navigate.

See comment in the event binding.

File: src/ui/components/search/autocompletecomponent.js [hunk 16]
```diff
@@ -145,22 +151,26 @@ export default class AutoCompleteComponent extends Component {
...
+        this.core.autoCompleteFilter(queryInput.value, this._experienceKey, this._barKey);
+      } else {
+        this.core.autoComplete(queryInput.value, this._experienceKey, this._barKey);
+      }
     });
 
     // Allow the user to navigate between the results using the keyboard
     DOM.on(queryInput, 'keydown', (e) => {
-     this.handleNavigateResults(e.keyCode, e);
+      this.handleNavigateResults(e.keyCode, queryInput.value, e);
```
Comment (b1lly) [inline src/ui/components/search/autocompletecomponent.js]: this.handleNavigateResults(e.keyCode, e);
this.handleSubmitResult(e.keyCode, queryInput.value);

It's possible to even check the keycode here, instead of within the handlers but I don't feel strongly.

File: src/ui/components/search/autocompletecomponent.js (hunk 11 above)
Comment (b1lly) [reply src/ui/components/search/autocompletecomponent.js]: Whatever is creating the data object that's being rendered into the markup should handle this, no?

File: src/ui/components/search/autocompletecomponent.js (hunk 9 above)
Comment (b1lly) [reply src/ui/components/search/autocompletecomponent.js]: I'd make it a separate handler. See comment above.

File: src/ui/components/search/filterpicker.js (hunk 12 above)
Comment (b1lly) [reply src/ui/components/search/filterpicker.js:4]: Yeah, but I also think some of the behavior is common between the both, not just the state. 

Like for instance, the browser history logic, and most of the search logic. I think we want to figure out a way to share the code in this regard, whichever direction we decide.

File: src/ui/components/search/filterpicker.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+     * @type {string} CSS selector
+     */
+    this._inputEl = opts.inputEl || '.js-yext-query';
+
+    /**
+     * The title used, provided to the template as a data point
+     * Optionally provided.
+     * @type {string}
+     */
+    this.title = opts.title;
```
Comment (b1lly) [inline src/ui/components/search/filterpicker.js:45]: Not sure this is used.

File: src/ui/components/search/filterpicker.js (hunk 14 above)
Comment (Delerme) [reply src/ui/components/search/filterpicker.js]: Yea, I guess it's more of the dropdown select vs text input bar with autocomplete. At the very least, I think associating this component with SearchBar is best to avoid confusion, so FilterBar seems better

File: src/ui/components/search/filterpicker.js (hunk 17 above)
Comment (Delerme) [reply src/ui/components/search/filterpicker.js:45]: Gets rendered above the search bar

File: src/core/models/filtersdata.js (hunk 15 above)
Comment (Delerme) [reply src/core/models/filtersdata.js]: Just filters, but they are similar enough it can be both. Made it a little more generic so now it can be used for both

File: src/ui/components/search/autocompletecomponent.js (hunk 16 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: done

File: src/ui/components/search/autocompletecomponent.js (hunk 11 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: yea done

File: src/ui/components/search/autocompletecomponent.js (hunk 9 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js]: done

File: src/ui/components/search/autocompletecomponent.js (hunk 8 above)
Comment (Delerme) [reply src/ui/components/search/autocompletecomponent.js:327]: done

File: src/core/models/autocompletedata.js [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+      if (j === matchedSubstrings.length - 1 && end < value.length) {
+        highlightedValue += value.slice(end);
+      }
+
+      nextStart = end;
+    }
+
+    return highlightedValue;
+  }
+}
```
Comment (b1lly) [inline src/core/models/autocompletedata.js:74]: Great! We might want to add tests for this. You can probably do it separate commit, if you want.

File: src/ui/components/search/filterpicker.js (hunk 14 above)
Comment (b1lly) [reply src/ui/components/search/filterpicker.js]: Per our discussion, lets leave it as FilterPicker or FilterSelector for now. And we can revisit it later.

**Review by b1lly — COMMENTED** [review]
LGTM minor comments ;) 

Good shit man.

File: src/core/models/autocompletedata.js (hunk 18 above)
Comment (Delerme) [reply src/core/models/autocompletedata.js:74]: Yea will do separately

**Comment by Delerme** [comment]
Switched the name to FilterPicker. Fun fact, you still need to override even if you want it to default, because the default is actually "Component" (since the type is set in Component's constructor). May want to fix that later, but for now doing the simple override

**Comment by b1lly** [comment]
LGTM
